Documentation and schema updates

GernotMaier opened the pull request:
This PR includes:

- significant improvement of main README.md
- documentation of schema for parameter and input data [schemas/README.md](schemas/README.md)
- simplified parameters schema as discussed in [gitlab pr 22](https://gitlab.cta-observatory.org/cta-computing/dpps/simulationpipeline/prototypes/notes/simulation-project-parameter-workflows/-/merge_requests/22). Note that the schema description is not 100% complete, but is thought to give a guidance on the approach.
- adaption of example schema files to this simplified description

VictorBarbosaMartins commented on schemas/MST_mirror_2f_measurements.schema.yml:
```diff
@@ -1,44 +1,39 @@
+        type: double
+      - name: psf_opt
+        description: |-
+          Spot size of mirror panel PSF
```
please refer in the description also what "opt" means.

GernotMaier replied:
This is a first example, let's move the discussion on column names somewhere else. This is what we have used in the whole simtelarray/model validation until now and is a first example.

VictorBarbosaMartins commented on schemas/README.md:
```diff
@@ -1,13 +1,122 @@
+The *data* field is used to describe the actual type, format, and allowed values of the parameter.
+
+- `type`: Data type of the parameter.
+- `units`: Units of the parameter (compatible with astropy units). Note that units are explicitly listed in the jsonschema.yml file.
```
what does compatible with astropy units mean? Is it that if you pass it as an argument to`astropy.Units()` it will give you an instance of `astropy.Quantity` ? If yes, then please include this explanation.

GernotMaier replied:
It means that astropy units can use these units. How and which functions are called is not so important in this context - isn't `compatible` saying exactly this?

VictorBarbosaMartins commented on schemas/README.md:
```diff
@@ -1,13 +1,122 @@
+      - name: photo_detection_efficiency
+        description: Average quantum or photon detection efficiency.
+        required: true
+        units: dimensionless
```
Related to the previous comment: I know that you prefer it to be more human readable, but note that if this is supposed to be astropy compatible, I think this should be`!astropy.units.Unit {unit: ''}`. (the same applies for all the units)

VictorBarbosaMartins replied:
Ah, now I see that the allowed units are also listed in `schemas/jsonschema.yml`. Please mention that somewhere in the text as well.

GernotMaier replied:
But astropy understands "km" or "dimensionless_unscaled" (see discussion #9). I really don't see a reason to make it look more complicated than it has to.

The comment that units a required to be listed (this is not a choice; json-schema requires it) is mentioned further down.

VictorBarbosaMartins reviewed (changes requested):
Hi @GernotMaier. Thanks for updating the documentation. Here I provide my comments, which consist of some suggestions to improve the text and mostly by questions, since I am not very used to the workflows. I will wait for some changes to approve, but this shouldnt take long I think. Since the your updates are only related to the documentation part I did not test the workflows. By the way, shouldnt we create tests for the workflows?
Thanks!

GernotMaier commented on README.md:
```diff
@@ -2,53 +2,92 @@
-- workflows for setting and validation of parameters are in [./workflows](./workflows).
-- command line tools used in these workflows and calling simtools applications are in [./workflows/tools](./workflows/tools).
-- input and generated data is validated using schemas located in [./schemas](./schemas).
+Workflows are used to set, derive, and validate simulation model parameters following all or a subset of these steps:
```
I don't think we want to use the cwl mechanism for the simulation production. Of course these are workflows, but we have dirac for that.

VictorBarbosaMartins replied:
Ok. So even if we use dirac for the other workflows, the dirac-kind of workflows are not going to be written here?

GernotMaier commented on README.md:
```diff
@@ -2,53 +2,92 @@
+Workflows are used to set, derive, and validate simulation model parameters following all or a subset of these steps:
 
-## Getting started
+Setting Workflows:
```
Wait - this is *setting a parameter*: you receive some data (need to define an API), assert the input data (e.g., are units/types are correct), derive the model parameters (e.g., a coordinate transformation or PSF simulations), validate that the outcome is consistent, and then review everything before accepting it. We have used this expression for a long time and and most documents (e.g., the concept document), my suggestion is not to change it.

VictorBarbosaMartins replied:
Ok! I had not understood it before but it is clear now, thanks.

VictorBarbosaMartins replied:
I think a good solution is to put "The" before "Setting", then it is clearer that we are talking about a type of workflows and about the procedure to set a workflow.

GernotMaier replied:
Let's not discuss if we have to add 'The' or not - for me this is an additional word without information.

GernotMaier commented on README.md:
````diff
@@ -2,53 +2,92 @@
 
-Install dependencies using mamba or conda:
-```
+Acceptance Workflows:
````
These are workflows which are part of the acceptance process: someone or machine has to look at the validation / data assertion output from the setting workflow; if successful we update the model database, and everything has to be documented. I like the bullet points, as it amplifies the different steps.

GernotMaier commented on README.md:
```diff
@@ -2,53 +2,92 @@
+
+- **workflows** encoded in [Common Workflow Language (CWL)](https://www.commonwl.org/) (see [./workflows](./workflows) in this repository)
+- **parameter schemas** defining all input data and simulation model parameters (see [./schemas](./schemas) in this repository)
+- **simulation tools and software** consisting of [simtools](https://github.com/gammasim/simtools) and the simulation software (e.g., [CORSIKA](https://www.iap.kit.edu/corsika/) and [sim_telarray](https://www.mpi-hd.mpg.de/hfm/~bernlohr/sim_telarray/)).
```
Good point, I've changed the sentence so that it is clearer that simtools are not part of the workflows. For containers: I seem them now as part of simtools, so not really necessary to mention them again.

GernotMaier commented on README.md:
```diff
@@ -2,53 +2,92 @@
+
+CWL workflows consists of
+
+- *Tools* are called by *steps* in a *Workflow* and are doing one single task. In most cases, tools are calling an application of [simtools](https://github.com/gammasim/simtools) including all required configuration parameters (e.g., [workflows/tools/derive_array_elements_coordinates.cwl](workflows/tools/derive_array_elements_coordinates.cwl)).
```
Hmm, not sure - my feeling is that it is the same, just with adding `: they`. Let's not get hold up on wording.

GernotMaier commented on README.md:
```diff
@@ -2,53 +2,92 @@
+CWL workflows consists of
+
+- *Tools* are called by *steps* in a *Workflow* and are doing one single task. In most cases, tools are calling an application of [simtools](https://github.com/gammasim/simtools) including all required configuration parameters (e.g., [workflows/tools/derive_array_elements_coordinates.cwl](workflows/tools/derive_array_elements_coordinates.cwl)).
+- *Workflows* connect *tools*, and allow to execute the steps discussed above (receive, assert, derive, validate, ...), see.e.g, workflow/DeriveArrayElementCoordinates.cwl](./workflows/DeriveArrayElementCoordinates.cwl).
```
Same comment as above, this too wordy.

GernotMaier commented on README.md:
```diff
@@ -2,53 +2,92 @@
+
 - assert input data
 - convert coordinates
 - assert derive parameter values
```
Yes.

GernotMaier commented on README.md:
```diff
@@ -2,53 +2,92 @@
+
+### Example workflow
 
 The following command line tool converts telescope coordinates from UTM to CORSIKA coordinates.
```
The coordinates are not defined in the yml file, but the configuration is in that file. I've added a sentence to make this clearer.

GernotMaier commented on README.md:
```diff
@@ -2,53 +2,92 @@
+- log files (stdout and stderr; ascii format)
 
-Use `cwltool --validate file_name.cwl` to check a workflow or command line tool for valid CWL.
+### CWL validation
```
If changed the sentence to say  `check...for valid CWL code`. This is like a linter and is not really connected the the assert steps (assert does more than validate).

GernotMaier commented on README.md:
```diff
@@ -2,53 +2,92 @@
+### CWL validation
 
-## Graphs
+Use `cwltool --validate file_name.cwl` to check a workflow file or command line tool to valid CWL code (this is also done by the CI).
```
It is to check for `valid CWL code`. It doesn't check the workflow, it does check that keywords are used correctly and all required keywords are in the cwl file.

GernotMaier commented on README.md:
```diff
@@ -2,53 +2,92 @@
-## Graphs
+Use `cwltool --validate file_name.cwl` to check a workflow file or command line tool to valid CWL code (this is also done by the CI).
+
+### Workflow Graphs
```
As discussed, this is part of the code organisation / validation, not of the workflow validation.

VictorBarbosaMartins replied:
Ok, maybe you show me one of the graphs next week again so that I get a better understanding of what they are. I tried to open the link with the example you provided in the text and got an error: `Error: The workflow could not be retrieved from the Git repository using the details given`

GernotMaier replied:
Fixed the path to the graph.

GernotMaier commented on README.md:
```diff
@@ -2,53 +2,92 @@
+
+### Workflow Graphs
 
 Prepare a workflow graph, e.g.:
```
see comment above.

GernotMaier commented on README.md:
````diff
@@ -2,53 +2,92 @@
 cwltool --print-dot DeriveArrayElementCoordinates.cwl | dot -Tsvg > DeriveArrayElementCoordinates.cwl.svg
 ```
 
 Alternatively, use https://view.commonwl.org/ , e.g., see [this example](https://view.commonwl.org/workflows/github.com/gammasim/workflows/blob/prototype-DeriveArrayElementCoordinates/workflows/DeriveArrayElementCoordinates.cwl).
````
Not sure what this comments applies to - `use it to what` means?

VictorBarbosaMartins replied:
Ah, sorry, probably I placed the comment on the wrong place.

GernotMaier commented on README.md:
```diff
@@ -2,53 +2,92 @@
 
+## Parameter schema files
+
+The parameter schema files are used to describe all input data and simulation model parameters and can be found in [./schemas](./schemas) in this repository.
```
Yes. Every model parameter has a schema. This is for the parameter definition, not for the values. Values are stored in the database, but not the definition.

GernotMaier commented on schemas/MST_mirror_2f_measurements.schema.yml:
```diff
@@ -1,44 +1,39 @@
+%YAML 1.2
```
Yes, this is a schema.

GernotMaier commented:
Thanks for the review - I think I've addressed all comments.

Most importantly, I've re-organized the README.md so that it is clearer what part is for users and what is for developers.

Let me know if this is good.

VictorBarbosaMartins reviewed (approved):
Hi @GernotMaier. Thanks for updating the docs and for the explanations before! It is clearer now. I just left two conversations open with suggestions but I am fine with all the changes.

GernotMaier commented:
@VictorBarbosaMartins - thanks for the review! I've fixed the path and will merge now.
